File: python/app/rbac.py
"""Server-side mirror of ../web/src/lib/rbac.ts.

Unlike a client-side SPA, this app's role checks *are* part of the security
boundary that stands between an end user and Mudbase — there is no browser
JS to "just hide a button" here, every write goes through one of these
gates before this app's own request handler ever calls Mudbase. That said,
Mudbase's own collection permissions remain the ultimate enforcement layer
(see plan/build-plan.md "Live smoke test results" for a raw write attempt
that bypasses this app's routers entirely) — these functions are this
app's own defense-in-depth check, not a replacement for it.
"""

_MANAGE_LISTS_ROLES = frozenset({"owner"})
_MANAGE_CARDS_ROLES = frozenset({"owner", "member"})
_KNOWN_ROLES = frozenset({"owner", "member", "viewer"})

_ROLE_LABELS = {"owner": "Owner", "member": "Member", "viewer": "Viewer"}


def can_manage_lists(role: str | None) -> bool:
    return role in _MANAGE_LISTS_ROLES


def can_manage_cards(role: str | None) -> bool:
    return role in _MANAGE_CARDS_ROLES


def is_read_only(role: str | None) -> bool:
    return not can_manage_cards(role)


def is_known_role(role: str | None) -> bool:
    return role in _KNOWN_ROLES


def role_label(role: str | None) -> str:
    return _ROLE_LABELS.get(role or "", "Unknown")

File: go/internal/rbac/rbac.go
// Package rbac defines this app's three role slugs and the operations each one may perform. This
// is a server-side mirror of the reference web app's src/lib/rbac.ts, used two ways in this app:
// (1) internal/server/middleware.go's requireRole gates every write route before a handler ever
// runs (the "server-side enforcement" the task requires, not just hidden template controls), and
// (2) the page templates read the same CanManage*/IsReadOnly flags to decide what to render.
// Neither of those is the real security boundary - Mudbase's own collection permissions are (see
// mbase.IsForbidden) - this package exists purely for defense-in-depth and correct UX, exactly as
// documented in plan/build-plan.md's "RBAC Matrix".
package rbac

// Role slugs this project's Multi-Role auth is configured with.
const (
	RoleOwner  = "owner"
	RoleMember = "member"
	RoleViewer = "viewer"
)

// CanManageLists reports whether role may create/rename/delete/reorder board columns.
func CanManageLists(role string) bool {
	return role == RoleOwner
}

// CanManageCards reports whether role may create/edit/delete/move cards.
func CanManageCards(role string) bool {
	return role == RoleOwner || role == RoleMember
}

// IsReadOnly reports whether role (or no role at all) can only ever read the board.
func IsReadOnly(role string) bool {
	return role == RoleViewer || role == ""
}

// Label renders a role slug for display, "Unknown" for anything unrecognized (including "").
func Label(role string) string {
	switch role {
	case RoleOwner:
		return "Owner"
	case RoleMember:
		return "Member"
	case RoleViewer:
		return "Viewer"
	default:
		return "Unknown"
	}
}

// IsValid reports whether role is one of this app's three known slugs.
func IsValid(role string) bool {
	return role == RoleOwner || role == RoleMember || role == RoleViewer
}
